docs/UPGRADE-20.0.md

```diff
@@ -0,0 +1,91 @@
+# Upgrade from v19.x to v20.x
+
+Please consult the `examples` directory for reference example configurations. If you find a bug, please open an issue with supporting configuration to reproduce.
+
+## List of backwards incompatible changes
+
+- Minium supported AWS provider version increased to `v5.0`
+- Minimum supported Terraform version increased to `v1.1` to support Terraform state `moved` blocks
+- The `resolve_conflicts` argument within the `cluster_addons` configuration has been replaced with `resolve_conflicts_on_create` and `resolve_conflicts_on_delete` now that `resolve_conflicts` is deprecated
+- The `cluster_addons` `preserve` argument default/fallback value is now set to `true`. This has shown to be useful for users deprovisioning clusters while avoiding the situation where the CNI is deleted too early and causes resources to be left orphaned which results in conflicts.
+- The Karpenter sub-module's use of the `irsa` naming convention has been replaced with `pod-identity` along with an update to the Karpenter controller IAM policy to align with the `v1beta1`/`v0.32` changes
```

Comment:
for those of us that want to keep using irsa, and not have pod-identity in the cluster, will there be an option to do so?

Reply:
I have a similar question. This module should continue to support IRSA

Reply:
I don't follow, can you elaborate? What is the use case or problem that you are solving for?

Reply:
we would like to continue using regular tradicional IRSA instead of moving to https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html using
```
"Principal": {
                "Service": "pods.eks.amazonaws.com"
            }
```

Reply:
For what though?

Reply:
for karpenter pods, for the way karpenter auths against AWS IAM
if the language is changing here to refer to `pod-identity`, it does indicate it will use a different method of talking to IAM

Reply:
Ah ok, yes - the Karpenter sub-module here will be using pod identify going forward starting in v20. The IAM policy is still TBD, but the intent is to follow pod identity

Reply:
can both not be supported?
that's the ask here

Reply:
No. And most likely the role will be removed from this sub-module in favor of consolidating on the in-progress https://github.com/clowdhaus/terraform-aws-eks-pod-identity

Why: pod identity allows for roles to be shared while still permitting scoped permissions based on session tags. The instance termination resources (queue, event rules, etc.) are still cluster specific and follow a 1:1 mapping.

Reply:
This relates to my question that was marked resolved: https://github.com/terraform-aws-modules/terraform-aws-eks/pull/2858#discussion_r1434694155

Pod-Identity Requires another DaemonSet which, amongst other things I am sure I am missing, is more Node overhead and security reviews(for some) to even be able to use. Locking a Karpenter Update(v0.32+) behind an authentication change with no overlap seems like it would do the opposite of drive the adoption of this module.

Since Karpenter supports both methods it would stand to reason that the preferred module would also support both?

Reply:
IRSA does not support session tags so you cannot have both for this particular policy. Pod identity is the successor to IRSA and that is the direction we are moving - you can continue to use IRSA on the v19.x module versions

Reply:
we can keep using v19, but we wont have updated iam policies for karpenter v.32+, correct?
or were all the require changes done in that version?

Reply:
Again, we need to be specific - what do you mean by "updated policies"? IRSA does not support session tags, so you cannot have a policy crafted for scoped re-use with session tags that pod identity supports and simply swap out the trust policy for IRSA

Reply:
Few things, not everyone wants to share Policies between Clusters for a myriad of reasons, both directions are valid. The issue here is that currently, v19.x, the Karpernter Policy is incompatible with the new CRDs provided in preparation for 1.x of Karpenter. This is a problem for many trying to stay on top of updates etc. The Issues requesting v0.32+ support indicate that this is to be solved in v20.x however this comes at the cost of also having to support Pod Identity which for some will be a blocker.

So if I am understanding this correctly it sounds like for anyone that does not want to use Pod Identity for any reason(for Karpenter in this case) then they have until v20.x to fork and migrate off of this module and unblock themselves with reg. to a Karpenter Update.

They could also stop using the Karpenter Sub-Module but I feel that with a sudden change like this teams would question the sustainability of the core module as well.

Reply:
Is this entire module removing support for IRSA? Unfortunately pod Identity is not supported in AWS GovCloud regions: https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html#pod-id-considerations

It’s a shame to me to lose out to other module updates due to this hard switch to pod identities

Reply:
> share Policies between Clusters for a myriad of reasons

Reasons, such as? This is a common pattern, no? Service linked roles, the role used by Terraform to execute applies, etc. I'm not sure how this applies to the direction of Pod Identity (just because you can use a common role doesn't dictate that you have to - we tend to want to move in the direction of more functionality and flexibility, not less - no?)

> The issue here is that currently, v19.x, the Karpernter Policy is incompatible with the new CRDs provided in preparation for 1.x of Karpenter.

Is this true? We have a number of folks using this sub-module with the latest Karpenter version today

> This is a problem for many trying to stay on top of updates etc. 

How does this relate to this module? This module is trying to move forward to support the latest features and functionality provided by Amazon EKS. Users are free to update/upgrade as they see fit, but this project is not held back by the pace at which users update/upgrade.

I feel like we're just fighting change here for the sake of fighting change. If you do not wish to use pod identity, you do not have to - nobody is forcing you. However, we are not going to hold back the development of this module because a few folks do not want to change. We can't satisfy all users - we still have users using very old versions of this module and thats perfectly fine. But this module will always strive to move forward with the latest features and advancements provided by Amazon EKS. 

Lastly, if this is all in reference to just the Karpenter sub-module, I would suggest you turn your attention here https://github.com/aws/containers-roadmap/issues/1792

Reply:
> Is this entire module removing support for IRSA? Unfortunately pod Identity is not supported in AWS GovCloud regions: [docs.aws.amazon.com/eks/latest/userguide/pod-identities.html#pod-id-considerations](https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html#pod-id-considerations)
> It’s a shame to me to lose out to other module updates due to this hard switch to pod identities

Again, please please please be specific. The root module will still, and as long as EKS supports it, and always support creating an OIDC identity provider which is used for IRSA authentication. Thats staying and not going anywhere, unless EKS drops support for that

If we are just talking about Karpenter, I will revert to say - the Karpenter sub-module current implementation is all TBD currently. Will it or will it not create an IAM role - I don't know. Does that prevent you from using this module?  Again, I don't know without more context. You can use the v19 version which does work

Reply:
> Lastly, if this is all in reference to just the Karpenter sub-module, I would suggest you turn your attention here https://github.com/aws/containers-roadmap/issues/1792
sure we all want that.
but until it's actually here, we still have to manage it ourselves

Reply:
> Reasons, such as? This is a common pattern, no? Service linked roles, the role used by Terraform to execute applies, etc. I'm not sure how this applies to the direction of Pod Identity (just because you can use a common role doesn't dictate that you have to - we tend to want to move in the direction of more functionality and flexibility, not less - no?)

Like I said, both patterns are valid and no one is debating that. I feel like flexibility would be supporting both authentication mechanisms but perhaps I am wrong.

> Is this true? We have a number of folks using this sub-module with the latest Karpenter version today

Based on #2733 it is hard to tell and we felt that waiting for v20.x made the most sense until now. We will happily test and confirm either way.

> How does this relate to this module? This module is trying to move forward to support the latest features and functionality provided by Amazon EKS. Users are free to update/upgrade as they see fit, but this project is not held back by the pace at which users update/upgrade.

If we use the module to create the IAM Policy that Karpenter is using and it is not compatible with the newer CRDs than we cannot update Karpenter no? That would mean that this module becomes blocking based on that functionality. You indicated that v19.x works with the new CRDs so again, we will happily test to confirm. No one is asking to hold the module back, we are just asking for consideration with regard to how changes are rolled out and impact the end user.

> I feel like we're just fighting change here for the sake of fighting change. If you do not wish to use pod identity, you do not have to - nobody is forcing you. However, we are not going to hold back the development of this module because a few folks do not want to change. We can't satisfy all users - we still have users using very old versions of this module and thats perfectly fine. But this module will always strive to move forward with the latest features and advancements provided by Amazon EKS.

I am sorry if you feel I am fighting a change just to fight a change, as someone who has been happily using this module for years now I am just fighting against breaking the ability for consumers to use the new latest features and advancements provided by the maintainers. If we want to fight a change just to fight a change we can argue about the arbitrary tag that is added in this PR that I still do not understand but we are not, we are discussing a functionality change that impacts the ability for some to use this module.

That being said, if we wish to continue you to use the Karpenter Sub-Module, in its projected state, then technically you are forcing us to either use Pod Identity or change the way in which we manage Karpenter or consume this module in general.

> Lastly, if this is all in reference to just the Karpenter sub-module, I would suggest you turn your attention here https://github.com/aws/containers-roadmap/issues/1792

As it was stated already, until this lands we manage it ourselves so it does not solve this.

Reply:
Pod identity is not supported on Fargate, so if using Karpenter with controllers on Fargate, need to continue using IRSA

Reply:
in v20.x, both pod identity and irsa will be supported - we don't have clear guidance yet on a policy with session tags so the policy will match the upstream project (as its currently written) and I'll keep working with the service teams on different avenues of improving policy support. 

This is one example I am exploring where the module would pull the policy from the projects repository based on the project's version:
- [Example of a project that contains the policy that will be pulled directly from the source](https://github.com/clowdhaus/example-external-policies)
- [Example of consuming that policy in a module that aligns the policy to the projects version](https://github.com/clowdhaus/example-external-policies-consumer)

IAM policies do not have a concept of versions, but with these projects that have a clear and direct link between their API functionality and the permissions required, I firmly believe there needs to be a direct link between the version of the software deployed and the version of the policy used. So if we take this situation where there were a number of API changes and permission changes going from Karpenter 0.28 to say 0.31 or later, when you update your Karpenter controller version, you would be updating the policy in lock-step. You could upgrade/downgrade to any version you like, the the API and the permissions are kept in sync. We also experienced this with the AWS load balancer controller and I'm sure it will keep happening unless there is a way to synchronize these two artifacts